vcd/resource_generic_crud.go

```diff
@@ -0,0 +1,192 @@
+
+func createResource[O updateDeleter[O, I], I any](ctx context.Context, d *schema.ResourceData, meta interface{}, c crudConfig[O, I]) diag.Diagnostics {
+	t, err := c.getTypeFunc(d)
+	if err != nil {
+		return diag.Errorf("error getting %s type: %s", c.entityLabel, err)
+	}
+
+	vcdClient := meta.(*VCDClient)
+	err = execSchemaHook(vcdClient, d, c.preCreateHooks)
+	if err != nil {
+		return diag.Errorf("error executing pre-create %s hooks: %s", c.entityLabel, err)
+	}
+
+	createdEntity, err := c.createFunc(t)
+	if err != nil {
+		return diag.Errorf("error creating %s: %s", c.entityLabel, err)
+	}
+
+	err = c.stateStoreFunc(d, createdEntity)
+	if err != nil {
+		return diag.Errorf("error storing %s to state: %s", c.entityLabel, err)
+	}
+
+	return c.resourceReadFunc(ctx, d, meta)
+}
+
+func updateResource[O updateDeleter[O, I], I any](ctx context.Context, d *schema.ResourceData, meta interface{}, c crudConfig[O, I]) diag.Diagnostics {
+	t, err := c.getTypeFunc(d)
+	if err != nil {
+		return diag.Errorf("error getting %s type: %s", c.entityLabel, err)
+	}
+
+	retrievedEntity, err := c.getEntityFunc(d.Id())
+	if err != nil {
+		return diag.Errorf("error getting %s: %s", c.entityLabel, err)
+	}
+
+	_, err = retrievedEntity.Update(t)
+	if err != nil {
+		return diag.Errorf("error storing %s to state: %s", c.entityLabel, err)
+	}
+
+	return c.resourceReadFunc(ctx, d, meta)
+}
+
+func readResource[O updateDeleter[O, I], I any](_ context.Context, d *schema.ResourceData, _ interface{}, c crudConfig[O, I]) diag.Diagnostics {
+	retrievedEntity, err := c.getEntityFunc(d.Id())
+	if err != nil {
+		if govcd.ContainsNotFound(err) {
+			util.Logger.Printf("[DEBUG] entity '%s' with ID '%s' not found. Removing from state", c.entityLabel, d.Id())
+		}
+		return diag.Errorf("error getting %s: %s", c.entityLabel, err)
+	}
+
+	err = execEntityHook(retrievedEntity, c.readHooks)
+	if err != nil {
+		return diag.Errorf("error executing read %s hooks: %s", c.entityLabel, err)
+	}
+
+	err = c.stateStoreFunc(d, retrievedEntity)
+	if err != nil {
+		return diag.Errorf("error storing %s to state: %s", c.entityLabel, err)
+	}
+
+	return nil
+}
+
+func readDatasource[O updateDeleter[O, I], I any](_ context.Context, d *schema.ResourceData, _ interface{}, c crudConfig[O, I]) diag.Diagnostics {
+	entityName := d.Get("name").(string)
+	retrievedEntity, err := c.getEntityFunc(entityName)
+	if err != nil {
+		return diag.Errorf("error getting %s by Name '%s': %s", c.entityLabel, entityName, err)
+	}
+
+	err = c.stateStoreFunc(d, retrievedEntity)
+	if err != nil {
+		return diag.Errorf("error storing %s to state: %s", c.entityLabel, err)
+	}
+
+	return nil
+}
+
+func deleteResource[O updateDeleter[O, I], I any](_ context.Context, d *schema.ResourceData, _ interface{}, c crudConfig[O, I]) diag.Diagnostics {
+	retrievedEntity, err := c.getEntityFunc(d.Id())
+	if err != nil {
+		return diag.Errorf("error getting %s: %s", c.entityLabel, err)
```

Comment:
Doubt, should it behave like `readResource`? I mean, if the entity is not found, just ignore it and finish.

Reply:
I think we never did it for delete. The reason is - a resource always gets read before deletion and if it it doesn't exist - it will be removed from state in read. However - the good point is - if we decide to change such approach - it should be just one patch for many resources :)